### PR TITLE
diffing_with_keys.ml: heterogeneous typing

### PR DESCRIPTION
**Pull request by gasche**

### CODE REVIEW

File: utils/diffing_with_keys.ml [hunk 1]
```diff
@@ -57,19 +57,32 @@ module Swap = Map.Make(struct
...
-  let kx, ky = key (data x), key (data y) in
+type ('l, 'r) keys = ('l -> string) * ('r -> string)
+
+let swap_edge (type l r state)
+    (key : (l, r) keys)
+    (state : state)
+    (x : l with_pos)
+    (y : r with_pos)
+  : Swap.key
+    * (l with_pos * r with_pos, l with_pos * r with_pos, state) partial_edge
```
Comment (gasche) [inline utils/diffing_with_keys.ml:73]: Note: I am not sure whether the type of swaps should be `(l * r, l * r) partial_edge` or rather `(l * r, r * l) partial_erge`.

File: utils/diffing_with_keys.ml [hunk 2]
```diff
@@ -57,19 +57,32 @@ module Swap = Map.Make(struct
...
+=
+  let kx, ky = fst key (data x), snd key (data y) in
   if kx <= ky then
     (kx,ky), Left (pos x, state, (x,y))
   else
     (ky,kx), Right(pos x,state, (x,y))
```
Comment (gasche) [inline utils/diffing_with_keys.ml:79]: If we used `(l * r, r * l)`, we would need to swap `(x, y)` into `(y, x)` in the `Right` case here.

File: utils/diffing_with_keys.ml [hunk 3]
```diff
@@ -78,38 +91,74 @@ let add_edge ex ey = match ex, ey with
...
+    (x : l with_pos) (y : r with_pos)
+  : (string with_pos * string with_pos) option
+=
+  let kx, ky = fst key (data x), snd key (data y) in
+  let k = if kx <= ky then kx, ky else ky, kx in
+  match Swap.find_opt k swaps with
   | None | Some (Left _ | Right _)-> None
   | Some Both (state, (ll,lr),(rl,rr)) ->
-      match test state ll rr,  test state lr rl with
+      match test state ll rr, test state rl lr with
```
Comment (gasche) [inline utils/diffing_with_keys.ml:145]: There is a change here, that is mandated by the typing from the choice of using `(l * r, l * r) partial_edge` for swaps.

I believe that the change is not observable, due to the fact that the typeful information inside the `result` case is ignored (and likely to be uninteresting in the `Ok` case anyway, which is the one we care about here).

File: typing/includecore.ml [hunk 4]
```diff
@@ -545,7 +548,7 @@ module Variant_diffing = struct
               cd1.cd_res cd2.cd_res cd1.cd_args cd2.cd_args with
       | Some reason ->
           Error (Diffing_with_keys.Type {pos; got=cd1; expected=cd2; reason})
-      | None -> Ok (Ident.name cd1.cd_id)
+      | None -> Ok ()
```
Comment (gasche) [inline typing/includecore.ml:551]: I suspect that this was a leftover from a previous implementation approach, that was accepted for no good reason due to lax typing.

Comment (Octachron) [reply typing/includecore.ml:551]: Most probably.

**Comment by Octachron** [comment]
Why not? I am not completely convinced that this is simpler. (or that there is non-symmetric use cases) but this seems fine.

**Comment by gasche** [comment]
I have two arguments in favor of the change:
- the signatures are helpful to follow what's going on in this highly-parametric code (otherwise error messages are hell if you touch something)
- it caught a place where we had dead code, and one place where we code was fine but subtly asymmetric

**Comment by Octachron** [comment]
Superseded by #4

**Pull request closed without merging**